lib/course/data.dart

```diff
@@ -73,6 +71,28 @@ class Course implements Entity<Course> {
 
   final Collection<Lesson> lessons;
   final Collection<Lesson> visibleLessons;
+
+  @override
+  bool operator ==(Object other) =>
+      other is Course &&
+      id == other.id &&
+      createdAt == other.createdAt &&
+      updatedAt == other.updatedAt &&
+      name == other.name &&
+      description == other.description &&
+      teacherIds.deeplyEquals(other.teacherIds, unordered: true) &&
+      color == other.color &&
+      isArchived == other.isArchived;
+  @override
+  int get hashCode => hashList([
```

Comment:
Iterables are a bit more complicated than it seems:
Of course, `[1,2].hashCode != [1,2].hashCode` (most of the time), because two lists with the same elements aren't considered equal (`==`).
Using `dart:ui`'s `hashList` only works for ordered comparisons: `hashList([1,2]) == hashList([1,2])`, not for unordered ones like this one: `hashList([1,2]) != hashList([2,1])`.
Sorting the list items with a custom comparator that uses the `hashCode` could work. But in many performance critical sections, `hashCode` is assumed to be O(1), so I'm not sure we should do that.
That's why here I consciously opted for omitting `contents` from the `hashCode`. That means, two `Course`s that only differ from each other by having different teachers need to be compared manually with `==` to discover they're actually not equal. I think that's acceptable.

Reply:
Then we should also exclude `permissions` and `roleIds` from `User.hashCode`